EIPS/eip-5163.md

```diff
@@ -0,0 +1,95 @@
+
+```typescript
+type TxOptionsEnhancement = {
+  proposedContracts: {
+    [contractId: CrossChainAddress]: {
+      proposedName: string,
+      abi?: Abi,
+      contractData?: ContractData
+    }
+  }
+}
+```
+
+Wallets MUST store the recommending site & time along with this data for forensic purposes.
+The user SHOULD be able to adjust the proposed names in the confirmation, to avoid collisions, and help personalize their local context.
+Any subsequent confirmations that are displaying the proposed contracts' names SHOULD show the proposed names.
+  - A mouse hover may provide the metadata about when this name was proposed.
+Any subsequent transaction confirmations that are displaying transactions to the proposed contracts SHOULD use the proposed data to render those transactions as readably as possible.
+Additional optional parameters MAY be added to the `proposedContracts` options object to allow new kinds of enhancement in the future.
+
+## Rationale
+- We need a solution that can address two aspects of transaction interaction
+  - What you're interacting with
+  - How you're interacting with it
+- We want something we can get into production ASAP
+- We want a solution that is as complete as possible
+  - If used correctly, it should be able to totally eliminate the un-readability problems of transactions.
+- If this were adopted widespread, it could also reduce the need to rely on centralized indexes of on-chain assets in general, since users would have more complete local views of their own assets.
+
+## Backwards Compatibility
+Since this EIP involves a new optional parameter whose name is not used in any method, it should not have any issues with backwards compatibility.
+
+## Security Considerations
+This proposal's security depends on a user's first interaction with a contract being from a trustworthy site. If we assume that most of the time that a user is being phished, they are being robbed of an asset they previously acquired, then it should be safe to assume that there was a prior opportunity for the user to be introduced to reliable metadata from the source that they originally acquired the asset (the metadata is as trustworthy as the asset they have is authentic).
+
+Achieving that initial safety would require a widespread push for adoption, but since it's just about adding a new optional parameter and it keeps a site's users more resistant to phishing, I think this effort could be effective.
+
+Some users would not be protected by this: Users who acquired their assets previously, and do not have the relevant metadata for the assets they are holding. Those users would be at no greater risk than they are at today. One of these users being phished to a malicious site could be either presented with a malicious transaction or malicious metadata, but for a scammer the current behavior (transfer the asset) is simply fewer steps to success, so it seems unlikely this would be a desireable attack for phishers.
```

Comment:
```suggestion
Some users would not be protected by this: Users who acquired their assets previously, and do not have the relevant metadata for the assets they are holding. Those users would be at no greater risk than they are at today. One of these users being phished to a malicious site could be either presented with a malicious transaction or malicious metadata, but for a scammer the current behavior (transfer the asset) is simply fewer steps to success, so it seems unlikely this would be a desirable attack for phishers.
```